File: python/factory/UserFactory.py
from dataclasses import dataclass
from typing import Literal

class User():
    name: str
    age: int


@dataclass
class AdminUser(User):
    name: str
    age: int


@dataclass
class GuestUser(User):
    name: str
    age: int


@dataclass
class RegularUser(User):
    name: str
    age: int


UserType = Literal["admin", "guest", "regular"]

class UserFactory:

    @staticmethod
    def create_user(user_type: UserType, payload: dict) -> User:
        name = payload["name"]
        age = payload["age"]

        if user_type == "admin":
            return AdminUser(name, age)
        elif user_type == "guest":
            return GuestUser(name, age)
        elif user_type == "regular":
            return RegularUser(name, age)
        else:
            raise ValueError("Invalid user type")

File: ts/factory/UserFactory.ts
export interface User {
    name: string;
    age: number;
}

export class AdminUser implements User {
    constructor(public name: string, public age: number) {}
}

export class GuestUser implements User {
    constructor(public name: string, public age: number) {}
}

export class RegularUser implements User {
    constructor(public name: string, public age: number) {}
}

export class UserFactory{
    public static createUser(type: "admin" | "guest" | "regular", payload: User): User {
        switch (type) {
            case "admin":
                return new AdminUser(payload.name, payload.age);
            case "guest":
                return new GuestUser(payload.name, payload.age);
            case "regular":
                return new RegularUser(payload.name, payload.age);
            default:
                throw new Error("Invalid user type");
        }
    }
}